bootstrap.py

```diff
@@ -77,45 +80,23 @@ def region():
 
 def resource_tags():
     """ Returns a dictionary of all resource tags for the current instance """
-    import boto.ec2
-    api = boto.ec2.connect_to_region(region())
-    tags = api.get_all_tags(filters={'resource-id': instance_id()})
-    return {tag.name: tag.value for tag in tags}
-
-
-def security_groups():
-    """ Returns a list of sercurity groups for the current instance """
-    return instance_metadata('security-groups').split('\n')
-
-
-def infer_tags(security_group):
-    """ Attempts to infer tags from a security group name """
-    import re
-    matches = re.search(r'(?P<Project>[\w-]+)-(?P<Role>\w+)$', security_group)
-    return matches.groupdict()
-
-
-def implicit_tags():
-    """ Returns a list of tags inferred from security groups """
-    return [infer_tags(name) for name in security_groups()]
-
-
-def discover(trait):
-    """ Tries to find a trait in tags, makes a reasonable guess if it fails """
-    if trait in resource_tags():
-        return [resource_tags()[trait]]
-    else:
-        return [implicit_tags()[trait]]
+    try:
+        import boto.ec2
+        api = boto.ec2.connect_to_region(region())
+        tags = api.get_all_tags(filters={'resource-id': instance_id()})
+        return {tag.name: tag.value for tag in tags}
+    except:
+        pass
 
 
 def project_path():
     """ Returns the forge path for the discovered project """
-    return discover('Project')[0] + '/'
+    return detect('Project') + '/'
 
 
 def role_paths():
     """ Returns a list of forge paths for all discovered roles """
-    return [project_path() + role + '/' for role in discover('Role')]
+    return project_path() + detect('Role') + '/'
```

Comment:
This changes the signature of `role_paths()`, and breaks a significant feature of `forge`. Multiple roles may be assigned to a single instance (`managed` and `application` as a common example).

This case is one of the reasons why `detect()` and `discover()` both exist. Iterables in environment variables are tricksy like Hobbitses. `$IFS` in your shell may not be the same as in `subprocess.call()`, which is subtle and easily overlooked when troubleshooting.